apps/examples/araplayer/AudioWorker.cpp

```diff
@@ -0,0 +1,268 @@
+			dMap = ((AudioPDStream *)sIter->second)->getDownloaderMap();
+			break;
+		case STREAM_HTTP_HLS:
+		case STREAM_HTTP_DASH:
+			dMap = ((AudioHASStream *)sIter->second)->getDownloaderMap();
+			break;
+		default:
+			DEBUG_TRACE(LOG_INFO, "Unknow stream type [%d]", ((AudioStream *)sIter->second)->getStreamType());
+			continue;
+		}
+
+		dIter = dMap.find(handle);
+		if (dIter != dMap.end()) {
+			DEBUG_TRACE(LOG_INFO, "found handle\n");
+			return dIter->second;
+		}
+	}
+}
+
+void AudioWorker::measureBandwidth()
+{
+	std::map<pid_t, AudioStream *> sMap = mPlayer->getStreamMap();
+	std::map<pid_t, AudioStream *>::iterator sIter;
+	std::map<void *, AudioHttpDownloader *> dMap;
+	std::map<void *, AudioHttpDownloader *>::iterator dIter;
+
+	double val = 0;
+	double totalVal = 0;
+	int conns = 0;
+	int ret = 0;
+
+	if (mMeasureCnt % 5 == 0) {
+		for (sIter = sMap.begin(); sIter != sMap.end(); sIter++) {
+			switch (((AudioStream *)sIter->second)->getStreamType()) {
+			case STREAM_HTTP_PROGRESSIVE:
+				dMap = ((AudioPDStream *)sIter->second)->getDownloaderMap();
+				break;
+			case STREAM_HTTP_HLS:
+			case STREAM_HTTP_DASH:
+				dMap = ((AudioHASStream *)sIter->second)->getDownloaderMap();
+				break;
+			default:
+				DEBUG_TRACE(LOG_INFO, "Unknow stream type [%d]", ((AudioStream *)sIter->second)->getStreamType());
+				continue;
+			}
+
+			for (dIter = dMap.begin(); dIter != dMap.end(); dIter++) {
+				ret = curl_easy_getinfo((CURL *)dIter->first, CURLINFO_SPEED_DOWNLOAD, &val);
+				if ((ret == CURLE_OK) && (val > 0)) {
+					totalVal += val;
+					conns++;
+				}
+			}
+		}
+		mBandwidth = ((totalVal / 1024 / 1024) * 8) / conns;
+		//DEBUG_TRACE(LOG_INFO, "Currnet Download Speed : %0.3f Mbps, Connections(%d)", mBandwidth, conns);
+	}
+
+	mMeasureCnt++;
+}
+
+void AudioWorker::pollCurl(int msec)
+{
+	CURLMsg *msg;
+	CURLMcode mc;
+
+	int numfds, msgs_left, repeats = 0;
+
+	mc = curl_multi_perform((CURLM *)mCurl, &mCurlRunning);
+
+	if (mc == CURLM_OK)
+		mc = curl_multi_wait((CURLM *)mCurl, NULL, 0, msec, &numfds);
+
+	if (mc != CURLM_OK) {
+		//DEBUG_TRACE(LOG_ERR, "curl_multi_wait() failed, code %d", mc);
+		return;
+	}
+
+	if (!numfds) {
+		repeats++; /* count number of repeated zero numfds */
+		if (repeats > 1)
+			WAITMS(100); /* sleep 100 milliseconds */
+	} else
+		repeats = 0;
+
+	/* See how the transfers went */
+	while ((msg = curl_multi_info_read((CURLM *)mCurl, &msgs_left))) {
+		if (msg->msg == CURLMSG_DONE) {
+			AudioHttpDownloader *httpDownloader = static_cast<AudioHttpDownloader *>(findHttpDownloader((void *)msg->easy_handle));
+			if (httpDownloader) {
+				DEBUG_TRACE(LOG_INFO, "HTTP Downlaod Complete??? %d", mc);
+				httpDownloader->downloadComplete();
+			} else {
+				// error process...
+				DEBUG_TRACE(LOG_INFO, "curl_multi_remove_handle %d", mc);
+				curl_multi_remove_handle((CURLM *)mCurl, msg->easy_handle);
+			}
+		}
+	}
+	return;
+}
+
+bool AudioWorker::processLoop()
+{
+	if (mEnableCurl) {
```

Comment:
stream->doSomething();
- Local : do nothing.
- others : execute code here (but PD also need to measuer Bandwidth?)